packages/slate/src/core/normalize-node.ts

```diff
@@ -1,123 +1,134 @@
+import {
+  Editor,
+  Element,
+  Descendant,
+  Text,
+  Transforms,
+  Node,
+  Path,
+  Ancestor,
+} from '../interfaces'
 
 export const normalizeNode: WithEditorFirstArg<Editor['normalizeNode']> = (
   editor,
   entry,
   options
 ) => {
-  const [node, path] = entry
+  const [node, path] = entry as [{}, Path] // node is not yet normalized, treat as hostile
 
   // There are no core normalizations for text nodes.
   if (Text.isText(node)) {
     return
   }
 
-  // Ensure that block and inline nodes have at least one text child.
-  if (Element.isElement(node) && node.children.length === 0) {
-    const child = { text: '' }
-    Transforms.insertNodes(editor, child, {
-      at: path.concat(0),
-      voids: true,
-    })
-    return
+  if (!('children' in node)) {
+    // If the node is not a text node, and doesn't have a `children` field,
+    // then we have an invalid node that will upset slate.
+    //
+    // eg: `{ type: 'some_node' }`.
+    //
+    // To prevent slate from breaking, we can add the `children` field,
+    // and now that it is valid, we can to many more operations easily,
+    // such as extend normalizers to fix erronous structure.
+    ;(node as Element).children = []
   }
+  let element = node as Ancestor // we will have to refetch the element any time we modify its children since it clones to a new immutable reference when we do
 
-  // Determine whether the node should have block or inline children.
-  const shouldHaveInlines = Editor.isEditor(node)
-    ? false
-    : Element.isElement(node) &&
-      (editor.isInline(node) ||
-        node.children.length === 0 ||
-        Text.isText(node.children[0]) ||
-        editor.isInline(node.children[0]))
+  // Ensure that elements have at least one child.
+  if (element !== editor && element.children.length === 0) {
+    const child = { text: '' }
+    Transforms.insertNodes(editor, child, { at: path.concat(0), voids: true })
+    element = Node.get(editor, path) as Element
+  }
 
-  // Since we'll be applying operations while iterating, keep track of an
-  // index that accounts for any added/removed nodes.
-  let n = 0
+  // Determine whether the node should have only block or only inline children.
+  // - The editor should have only block children.
+  // - Inline elements should have only inline children.
+  // - Elements that begin with a text child or an inline element child should have only inline children.
+  // - All other elements should have only block children.
+  const shouldHaveInlines =
+    !(element === editor) &&
+    (editor.isInline(element) ||
+      Text.isText(element.children[0]) ||
+      editor.isInline(element.children[0]))
 
-  for (let i = 0; i < node.children.length; i++, n++) {
-    const currentNode = Node.get(editor, path)
-    if (Text.isText(currentNode)) continue
-    const child = currentNode.children[n] as Descendant
-    const prev = currentNode.children[n - 1] as Descendant
-    const isLast = i === node.children.length - 1
```

Comment:
old implementation used `i` as an index into the snapshot of the node children passed into `normalizeNode` and `n` as an index into the children of the live node as its changed--this would be useful if we could use `i` to avoid `Node.get` calls by reading from the cached `node` whenever possible, but needing to know the prev child, which might be a newly added node, creates complexity.